File: 2022/day6/src/main/java/Main.java
package main.java;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;

public class Main {
    private static final int MARKER_SIZE_1 = 4;
    private static final int MARKER_SIZE_2 = 14;
    public static void main(String[] args) throws IOException {
        var input = Files.lines(Paths.get("2022/day6/src/main/resources/input.txt")).toList();
        // var input = Arrays.asList(
        //         "bvwbjplbgvbhsrlpgdmjqwftvncz",
        //             "nppdvjthqldpwncqszvftbrmjlhg",
        //             "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg",
        //             "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw"
        //             );
        // var input = Arrays.asList(
        //         "mjqjpqmgbljsphdztnvjfqwrcgsmlb",
        //             "bvwbjplbgvbhsrlpgdmjqwftvncz",
        //             "nppdvjthqldpwncqszvftbrmjlhg",
        //             "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg",
        //             "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw"
        //             );

        input.forEach(i ->
            System.out.printf("%s, first marker: %d, first message: %d\n", 
                            i, nextMarkerSequence(i, 0, MARKER_SIZE_1), nextMarkerSequence(i, 0, MARKER_SIZE_2))
        );
    }

    private static int nextMarkerSequence(String input, int index, int markerSize) {
        for(; index < input.length() - markerSize; index++) {
            if (!hasDuplicate(input.substring(index, index + markerSize))) {
                return index + markerSize;
            }
        }
        return -1;
    }

    private static boolean hasDuplicate(String input) {
        for(int i = 0; i < input.length(); i++) {
            for(int c = i + 1; c < input.length(); c++) {
                if (input.charAt(i) == input.charAt(c)) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: 2022/day6/src/main.rs
use std::convert::TryInto;

fn detect_marker(input: &str, size: u32) -> u32 {
    let ns = size as usize;
    for index in 0..(input.len() - ns) {
        if !has_duplicate(&input[index..index+ns]) {
          return (index + ns).try_into().unwrap(); 
        }
    }
    return u32::MIN;
}

fn has_duplicate(part: &str) -> bool {
    for (i,left) in part.char_indices() {
        match part[i+1..].find(left) {
            None => {},
            Some(_) => return true,
        };
    }
    return false;
}


fn main() {
    let input = [
        "bvwbjplbgvbhsrlpgdmjqwftvncz", //: first marker after character 5
        "nppdvjthqldpwncqszvftbrmjlhg", //: first marker after character 6
        "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", //: first marker after character 10
        "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", //: first marker after character 11
        
        "mjqjpqmgbljsphdztnvjfqwrcgsmlb", // first marker after character 19
        "bvwbjplbgvbhsrlpgdmjqwftvncz", // first marker after character 23
        "nppdvjthqldpwncqszvftbrmjlhg", // first marker after character 23
        "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg", // first marker after character 29
        "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw", // first marker after character 26
    ];

    for i in &input[..4] {
        println!("text: {}, marker: {}", &i, detect_marker(i, 4));
    }
    for i in &input[4..] {
        println!("text: {}, marker: {}", &i, detect_marker(i, 14));
    }
}
